sldns/str2wire.c

```diff
@@ -615,6 +614,119 @@ sldns_affix_token(sldns_buffer* strbuf, char* token, size_t* token_len,
+	uint8_t* svcparams[64];
+	uint8_t* mandatory = NULL;
+	uint8_t* rdata_ptr = rdata;
+	uint16_t rdata_remaining = rdata_len;
+
+	/* find the SvcParams */
+	while (rdata_remaining) {
+		uint16_t svcbparam_len;
+
+		svcparams[nparams] = rdata_ptr;
+		if (rdata_remaining < 4)
+			return LDNS_WIREPARSE_ERR_SVCPARAM_BROKEN_RDATA;
+		svcbparam_len = sldns_read_uint16(rdata_ptr + 2);
+		rdata_remaining -= 4;
+		rdata_ptr += 4;
+
+		if (rdata_remaining < svcbparam_len)
+			return LDNS_WIREPARSE_ERR_SVCPARAM_BROKEN_RDATA;
+		rdata_remaining -= svcbparam_len;
+		rdata_ptr += svcbparam_len;
+
+		nparams += 1;
+		if (nparams > MAX_NUMBER_OF_SVCPARAMS)
+			return LDNS_WIREPARSE_ERR_SVCB_TOO_MANY_PARAMS;
+	}
+
+	/* In draft-ietf-dnsop-svcb-https-06 Section 7:
+	 *
+	 *     In wire format, the keys are represented by their numeric
+	 *     values in network byte order, concatenated in ascending order.
+	 */
+	qsort((void *)svcparams
+	     ,nparams
+	     ,sizeof(uint8_t*)
+	     ,sldns_str2wire_svcparam_key_cmp);
+
+
+	/* The code below revolves around sematic errors in the SVCParam set.
+	 * So long as we do not distinguish between running Unbound as a primary
+	 * or as a secondary, we default to secondary behavior and we ignore the
+	 * sematic errors. */
+
+	/* In draft-ietf-dnsop-svcb-https-06 Section 7:
+	 *
+	 *     Keys (...) MUST NOT appear more than once.
+	 *
+	 * If they key has already been seen, we have a duplicate
+	 */
+	// for (i = 0; i < nparams; i++) {
+	// 	uint16_t key = sldns_read_uint16(svcparams[i]);
+
+	// 	if (i + 1 < nparams && key == sldns_read_uint16(svcparams[i+1]))
+	// 		return LDNS_WIREPARSE_ERR_SVCB_DUPLICATE_KEYS;
+
+	// 	if (key == SVCB_KEY_MANDATORY)
+	// 		mandatory = svcparams[i];
+	// }
+
+	/* 4. verify that all the SvcParamKeys in mandatory are present */
+	// if (mandatory) {
+
+	// 	/* Divide by sizeof(uint16_t)*/
+	// 	uint16_t mandatory_nkeys = sldns_read_uint16(mandatory + 2) / sizeof(uint16_t);
+
+	// 	/* Guaranteed by sldns_str2wire_svcparam_key_value */
+	// 	assert(mandatory_nkeys > 0);
+
+	// 	for (i = 0; i < mandatory_nkeys; i++) {
+	// 		uint16_t mandatory_key = sldns_read_uint16(mandatory
+	// 		                       + 2 * sizeof(uint16_t)
+	// 		                       + i * sizeof(uint16_t));
+	// 		uint8_t found = 0;
+
+	// 		for (j = 0; j < nparams; j++) {
+	// 			if (mandatory_key == sldns_read_uint16(svcparams[j]))
+	// 				found = 1;
+	// 		}
+
+	// 		if (!found)
+	// 			return LDNS_WIREPARSE_ERR_SVCB_MANDATORY_MISSING_PARAM;
+	// 	}
+
+	// }
+
+	/* Write rdata in correct order */
+	// for (i = 0; i < nparams; i++) {
+	// 	uint16_t svcparam_len = sldns_read_uint16(svcparams[i] + 2)
+	// 	                      + 2 * sizeof(uint16_t);
+
+	// 	if (new_rdata_ptr + svcparam_len - new_rdata > sizeof(new_rdata))
+	// 		return LDNS_WIREPARSE_ERR_BUFFER_TOO_SMALL;
+
+	// 	memcpy(new_rdata_ptr, svcparams[i], svcparam_len);
+	// 	new_rdata_ptr += svcparam_len;
+	// }
+	// memcpy(rdata, new_rdata, rdata_len);
```

Comment:
Writing svcparams in sorted order is not part of semantic errors.
Also replaces `// ` style commenting with excluding by preprocessing statements
```suggestion
#if 0
	/* In draft-ietf-dnsop-svcb-https-06 Section 7:
	 *
	 *     Keys (...) MUST NOT appear more than once.
	 *
	 * If they key has already been seen, we have a duplicate
	 */
	for (i = 0; i < nparams; i++) {
		uint16_t key = sldns_read_uint16(svcparams[i]);

		if (i + 1 < nparams && key == sldns_read_uint16(svcparams[i+1]))
			return LDNS_WIREPARSE_ERR_SVCB_DUPLICATE_KEYS;

		if (key == SVCB_KEY_MANDATORY)
			mandatory = svcparams[i];
	}

	/* 4. verify that all the SvcParamKeys in mandatory are present */
	if (mandatory) {

		/* Divide by sizeof(uint16_t)*/
		uint16_t mandatory_nkeys = sldns_read_uint16(mandatory + 2) / sizeof(uint16_t);

		/* Guaranteed by sldns_str2wire_svcparam_key_value */
		assert(mandatory_nkeys > 0);

		for (i = 0; i < mandatory_nkeys; i++) {
			uint16_t mandatory_key = sldns_read_uint16(mandatory
			                       + 2 * sizeof(uint16_t)
			                       + i * sizeof(uint16_t));
			uint8_t found = 0;

			for (j = 0; j < nparams; j++) {
				if (mandatory_key == sldns_read_uint16(svcparams[j]))
					found = 1;
			}

			if (!found)
				return LDNS_WIREPARSE_ERR_SVCB_MANDATORY_MISSING_PARAM;
		}

	}
#endif
	/* Write rdata in correct order */
	for (i = 0; i < nparams; i++) {
		uint16_t svcparam_len = sldns_read_uint16(svcparams[i] + 2)
		                      + 2 * sizeof(uint16_t);

		if (new_rdata_ptr + svcparam_len - new_rdata > sizeof(new_rdata))
			return LDNS_WIREPARSE_ERR_BUFFER_TOO_SMALL;

		memcpy(new_rdata_ptr, svcparams[i], svcparam_len);
		new_rdata_ptr += svcparam_len;
	}
	memcpy(rdata, new_rdata, rdata_len);
```